lib/sources.nix

```diff
@@ -101,4 +101,201 @@ rec {
+        srcDirsTree = pathComponentsToTree (map splitPath srcDirs);
+      in
+        # The actual filter function with per-file processing
+        fullPath: type:
+          let
+            fileName = baseNameOf (toString fullPath);
+
+            components = splitPath fullPath;
+
+            isExplicitSrcFile = builtins.hasAttr fullPath srcFilesSet;
+            isExplicitSrcDir = type == "directory" && builtins.hasAttr fullPath srcDirsSet;
+            # The below is equivalent to
+            #   any (srcDir: hasPrefix (toString srcDir + "/") fullPath) srcDirs;
+            # but faster than O(n) where n is the number of `srcDirs` entries.
+            isUnderSomeSrcDir = isPrefixOfLeafPath components srcDirsTree;
+
+            isHidden = excludeHidden && hasPrefix "." fileName;
+
+            hasExcludedComponentInPath = any (c: elem c pathComponentExcludes) components;
+
+            isSourceInput =
+              isExplicitSrcFile ||
+              isExplicitSrcDir ||
+              (isUnderSomeSrcDir && !isHidden && !hasExcludedComponentInPath);
+
+            tracing =
+              let
+                prefix = if debugTracePrefix == "" then "" else debugTracePrefix + ": ";
+                action = if isSourceInput then "include" else "skip   ";
+                # Pad "regular" to be as wide as "directory" for aligned output.
+                formattedType = if type == "regular" then "regular  " else type;
+              in
+              if debugEnableTracing
+                then builtins.trace "${prefix}${action} ${formattedType} ${fullPath}"
+                else id;
+          in
+            tracing isSourceInput;
+
+  # A general-purpose, explicit source code importer suitable for most
+  # packaging needs.
+  #
+  # See `explicitSourceFilter` for an explanation of the filter arguments.
+  #
+  # Example usage:
+  #
+  #    src = lib.explicitFilterSource "mypackage" ./. {
+  #      srcDirs = [
+  #        ./src
+  #        ./app
+  #        ./images
+  #      ];
+  #      srcFiles = [
+  #        ./mypackage.cabal
+  #        ./Setup.hs
+  #      ];
+  #      pathComponentExcludes = [ "build" "gen" ];
+  #    };
+  #
+  # Note that `srcDirs = [ ./. ]` is also permitted.
+  #
+  # But consider that in most cases you will not want to include files
+  # that are not relevant for the build, such as `.nix` files, so that
+  # input hashes do not change unnecessarily.
+  #
+  # If you want to combine it with other filters, use `explicitSourceFilter`
+  # directly instead.
+  explicitFilterSource = name: topPath: filterArgs:
+    builtins.path {
+      # The `-src` suffix makes it easy to distinguish source store paths
+      # from built output store paths in the nix store.
+      #
+      # Requiring an explicit name prevents the basename of the directory
+      # making it into the store path when `./.` is used, thus preventing
+      # impure builds in situations where ./. is a directory with random
```

Comment:
Not sure how I feel about requiring a name, it makes the function more annoying to use, and I personally don't care much about being able to identify source derivations, so I'd be fine with some constant `name = "filtered-src"`. Having to specify a name for a `src` and the derivation itself is a bit weird too.

Reply:
I like how eg. `fetchpatch` has an optional name. imo a default constant is great though.

Reply:
OK; I think it's very useful to have good names in `nix-store -qR` or to identify slow downloads, but given that it being optional seems to be popular, so I will change it accordingly.

Luckily `fetchpatch` users have started to use `name` more aggressively; hopefully it will also be the case here.